### DISCUSSION TITLE
firefox freezes  a lot

### DISCUSSION THREAD

**Discussion by enaut** [ltsp/ltsp#173] [Q&A]
when loading up Firefox with ltsp 19 the startup is fast enough but as soon as some network traffic happens firefox freezes and freezes the whole system with it for quite some time. I did not find anything unusual in the logs and double checked that the memory is not the issue but right now this programm is not usable in gnome3 (wayland and nonwayland).

Any suggestions what I could do to locate the issue? I checked so far:
  * memory (RAM)
  * systemd journal
  * starting from terminal and looking for errors

client specs:
  * dualcore
  * 2-3GB ram (no difference between them)
  * ssh authentication and home mounting

With the old ltsp firefox was never fast but usable.

**Comment by alkisg** [comment]
Hmm, the only thing that changed there is the absence of "disable_hardlinks" in the SSHFS mount.
But if firefox was fighting with locks, it should show up in `top` at the moment of the lag.

Try having a terminal open with `top` running while that happens.
Or, try Ctrl+Alt+F1 when that happens, to see if at least you can change VTs.
Also, anything in dmesg or Xorg.log?

If those reveal nothing, it would be worth it to try with NFS home, temporarily, to see if SSHFS is to blame.

**Comment by enaut** [comment]
I tried a lot now and don'T seem to get any nearer to the problem.
* as the whole ui is frozen I can neither run top and see something nor switch to VT.
* Xorg.log doensnt exist.
* with nfs in the journal were some messages "nfs timeout still trying".
* with nfs it is slightly less and recovers quicker
* the whole ui is frozen but root local with epoptes from a different pc works
* top does show no high cpu usage (neither server nor client)
* on the client firefox has probably state D in top while frozen
* the problem is not only firefox it seems to be always when multiple IO happens at the same files. For example nautilus can crash too.

**Comment by alkisg** [comment]
Can you test with a VM client running on the LTSP server (or on the same host if it's already a VM), to make sure it's network-related?

**Comment by enaut** [comment]
Yes I will do that. I'm more and more suspecting that some network cable between server and clients is not that reliable... pings don't get lost but it seems as soon as there is traffic something gets wanky.

**Comment by enaut** [comment]
I checked the network... I did not find any issue (which does not mean there is none - I just did not find any)

I also tried launching firefox in private mode so that it does not require home directory information (or much less) however still firefox was freezing the whole UI for up to a minute. however it really does not seem to be firefox it's just that firefox triggers the behavior reliably. I guess its rather gnome shell or the graphics driver that freezes the UI.

Unfortunately I do not have time to test a lot further. I'm now trying to install mate and see what happens then.

**Comment by alkisg** [comment]
I proposed a VM client because it would make sure the issue wasn't related to network, flow control settings, lost packets, bad graphics card driver etc. It should be a quick test that answers a lot of questions.

**Comment by enaut** [comment]
Yes I know exactly why you suggested this. It's just there is no screen attached to the server and it's deep down in the basement. I have VNC access but this is not really good for testing as this stutters too. However I did not notice the 30secs+ freezes. Its just not a reliable outcome as the connection is not that good. And all the hardware-carrying-tests are hard to do in between lessons.

**Comment by enaut** [comment]
but thanks for your thoughts and help!

**Comment by alkisg** [comment]
@enaut, I'm thinking it might be related to NFS. I saw something similar today with one specific client that had a 100 Mbps NIC, instead of my usual testing with gigabit NICs.

I've the feeling it's related to either:
 - Mixed speed networks; somewhere gigabit, somewhere 100 Mbps,
 - The nfsmount klibc program that initramfs-tools uses to mount root,
 - Older versions of NFS that use UDP,
 - If UDP is involved, maybe disabling flow control there actually harms performance
 - Maybe even 32bit clients behave differently than 64bit clients (again due to NFS bugs)

My quick test was:

```shell
dd if=/dev/loop0 of=/dev/null bs=1M status=progress
```

...where loop0 was the automatically mounted root image, i.e. /srv/ltsp/images/x86_64.img.
And while normally this gives a constant transfer rate, with this specific client it lagged every few secs.

I'll do some more tests in the next days, and see if e.g. replacing `nfsmount` with `mount -t nfs` or using different NFS versions/options make the problem go away.

**Comment by alkisg** [comment]
These are the options that `nfsmount` uses:

> `# grep nfs /proc/self/mountinfo`
26 0 0:23 / /root ro,relatime - nfs 10.161.254.11:/srv/ltsp ro,vers=3,rsize=1048576,wsize=1048576,namlen=255,hard,nolock,proto=tcp,port=2049,timeo=7,retrans=10,sec=sys,local_lock=none,addr=10.161.254.11

And these are the options that `mount -t nfs` uses:

> `# grep nfs /proc/self/mountinfo`
365 29 0:55 / /home rw,relatime shared:226 - nfs4 server:/home/nfs rw,vers=4.2,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=10.161.254.33,local_lock=none,addr=10.161.254.11

...I don't see anything wrong, both are tcp; I'll test more when I'm there as I can't reproduce any issues with VMs etc remotely.

**Comment by alkisg** [comment]
First results show that the problem is in klibc nfsmount.
The same export gives 100 Mbps on the test client with `mount`,
and around 50 with `nfsmount`, but with extreme lags that last up to seconds.

Currently doing more tests.

**Comment by alkisg** [comment]
I opened https://github.com/ltsp/ltsp/issues/27 for the nfsmount issue.

If what you're experiencing is the same, let's continue there.
If your firefox issue is unrelated to nfsmount, let's continue that here.

**Comment by enaut** [comment]
Mhm I think I'm getting closer. However I'm still struggeling...

The problem is network related. I tried the VM you suggested which works and I tried to use a different network there is a students network and a teachernetwork going to the classroom.
The teachernetwork works flawless while the Studentsnetwork does not.

I did try to `scp ubuntu.iso remote://ubuntu.iso` the ubuntu image just 2GB of data in different directions:
#### teacher-net:
* server-> client  50MB/s
* client -> server 50MB/s
#### students-net:
* server -> client 50MB/s
* client -> server initially some data is sent (1000kB/s), then the speed is slower and slower and stops. The file never goes to the server. About 6 MB are sent according to the scp stats.

The strange thing is: ssh and ping work reliably in both directions.

So this is really a specific question to my setup I understand that you might not be able to guess whats wrong.
But you might aswell have seen such an issue before so here I am ;)

**Comment by alkisg** [comment]
Sure, enaut that's what the community issues are here for, to help each other even if the solution isn't part of the LTSP code base.

Answers to the following 2 questions would be most useful:
1) Is there any 100 Mbps connection from the server to the students-net, or is everything gigabit there?
2) Can you select e.g. 3 students-net clients in Epoptes, and execute a LAN benchmark from the menu, and paste the result here?

**Comment by enaut** [comment]
I can try to do that.
The lan benchmark is supposed to run on epoptes on the server i guess?(as I have configured epoptes to run on a computer in the classroom which is connected to the same switch)

The server is connected Gigabit to the Lab there is a switch with 2 Gigabit connections and the clients are connected to 100 Mbit ports.

However the scp tests were also done from a Gigabit connection.

I just created a testrun of iperf between server and client which shows good speed:
```
dietrich@Lehrer:~$ iperf -s                                                                                                                                                                                                   
------------------------------------------------------------                                                                                                                                                                  
Server listening on TCP port 5001                                                                                                                                                                                             
TCP window size: 85.3 KByte (default)                                                                                                                                                                                         
------------------------------------------------------------                                                                                                                                                                  
[  4] local 192.168.10.100 port 5001 connected with 192.168.10.53 port 52708                                                                                                                                                  
[ ID] Interval       Transfer     Bandwidth                                                                                                                                                                                   
[  4]  0.0-10.0 sec  1.10 GBytes   941 Mbits/sec                                                                                                                                                                              
^Cdietrich@Lehrer:~$ iperf -c 192.168.10.53                                                                                                                                                                                   
------------------------------------------------------------                                                                                                                                                                  
Client connecting to 192.168.10.53, TCP port 5001                                                                                                                                                                             
TCP window size: 85.0 KByte (default)                                                                                                                                                                                         
------------------------------------------------------------                                                                                                                                                                  
[  3] local 192.168.10.100 port 53880 connected with 192.168.10.53 port 5001                                                                                                                                                  
[ ID] Interval       Transfer     Bandwidth                                                                                                                                                                                   
[  3]  0.0-10.0 sec  1008 MBytes   846 Mbits/sec 
```

**Comment by alkisg** [comment]
Actually I just noticed this part better:

> client -> server initially some data is sent (1000kB/s), then the speed is slower and slower and stops. The file never goes to the server. About 6 MB are sent according to the scp stats.

I've never seen this; I only had problems in the opposite direction, server => clients.

I'd test what happens if you exchange a student computer with a teacher computer.
It would ensure that the problem is not related to the specific clients / network cards etc at all.

**Comment by enaut** [comment]
So I was going along the cable connection to see where the problem originates and then it was gone. I think it was a bad connected fibre optic cable.
